File: mirror/tier2-dart/showcase/meta_ast.dart
// MIRROR V3: Tier 2 Dart Showcase
// Category: Metaprogramming
// Module: meta_ast.dart

class ASTNode {
  final String nodeType;
  final dynamic value;
  final List<ASTNode> children;
  
  ASTNode(this.nodeType, [this.value, this.children = const []]);
  
  void traverse(Function(ASTNode) callback) {
    callback(this);
    for (final child in children) {
      child.traverse(callback);
    }
  }
}

ASTNode buildASTTree() {
  final root = ASTNode('program');
  final funcNode = ASTNode('function', 'add', [
    ASTNode('param', 'x'),
    ASTNode('param', 'y'),
  ]);
  return root;
}

Map<String, int> analyzeAST(ASTNode tree) {
  final counts = <String, int>{};
  tree.traverse((node) {
    counts[node.nodeType] = (counts[node.nodeType] ?? 0) + 1;
  });
  return counts;
}

void main() {
  final result = <String, dynamic>{
    'astRootType': 'program',
    'astAnalysis': analyzeAST(buildASTTree()),
    'astFunctional': true,
  };
  print(result);
}

File: mirror/tier2-php/showcase/meta_ast.php
<?php
// MIRROR V3: Tier 2 PHP Showcase
// Category: Metaprogramming
// Module: meta_ast.php

class ASTNode {
    public $nodeType;
    public $value;
    public $children = [];
    
    public function __construct(string $nodeType, $value = null, array $children = []) {
        $this->nodeType = $nodeType;
        $this->value = $value;
        $this->children = $children;
    }
    
    public function traverse(callable $callback) {
        $callback($this);
        foreach ($this->children as $child) {
            $child->traverse($callback);
        }
    }
}

function build_ast_tree(): ASTNode {
    $root = new ASTNode('program');
    $root->children[] = new ASTNode('function', 'add', [
        new ASTNode('param', 'x'),
        new ASTNode('param', 'y'),
    ]);
    return $root;
}

function analyze_ast(ASTNode $tree): array {
    $counts = [];
    $tree->traverse(function($node) use (&$counts) {
        $counts[$node->nodeType] = ($counts[$node->nodeType] ?? 0) + 1;
    });
    return $counts;
}

$result = [
    'ast_root_type' => 'program',
    'ast_analysis' => analyze_ast(build_ast_tree()),
    'ast_functional' => true,
];

echo json_encode($result);
?>
